esmvalcore/preprocessor/_multimodel.py

```diff
@@ -201,158 +185,102 @@ def _put_in_cube(template_cube, cube_data, statistic, t_axis):
+    return cf_units.Unit("days since 1850-01-01", calendar="standard")
 
-def _align_yearly_axes(cube):
-    """Perform a time-regridding operation to align time axes for yr data."""
-    years = [cell.point.year for cell in cube.coord('time').cells()]
-    # be extra sure that the first point is not in the previous year
-    if 0 not in np.diff(years):
-        return regrid_time(cube, 'yr')
-    return cube
 
-
-def _get_overlap(cubes):
-    """
-    Get discrete time overlaps.
-
-    This method gets the bounds of coord time
-    from the cube and assembles a continuous time
-    axis with smallest unit 1; then it finds the
-    overlaps by doing a 1-dim intersect;
-    takes the floor of first date and
-    ceil of last date.
+def _unify_time_coordinates(cubes):
     """
-    all_times = []
-    for cube in cubes:
-        span = _datetime_to_int_days(cube)
-        start, stop = span[0], span[-1]
-        all_times.append([start, stop])
-    bounds = [range(b[0], b[-1] + 1) for b in all_times]
-    time_pts = reduce(np.intersect1d, bounds)
-    if len(time_pts) > 1:
-        time_bounds_list = [time_pts[0], time_pts[-1]]
-        return time_bounds_list
+    Make sure all cubes' share the same time coordinate.
 
+    This function extracts the date information from the cube and
+    reconstructs the time coordinate, resetting the actual dates to the
+    15th of the month or 1st of july for yearly data (consistent with
+    `regrid_time`), so that there are no mismatches in the time arrays.
 
-def _slice_cube(cube, t_1, t_2):
-    """
-    Efficient slicer.
+    If cubes have different time units, it will use reset the calendar to
+    a default gregorian calendar with unit "days since 1850-01-01".
 
-    Simple cube data slicer on indices
-    of common time-data elements.
+    Might not work for (sub)daily data, because different calendars may have
+    different number of days in the year.
     """
-    time_pts = [t for t in cube.coord('time').points]
-    converted_t = _datetime_to_int_days(cube)
-    idxs = sorted([
-        time_pts.index(ii) for ii, jj in zip(time_pts, converted_t)
-        if t_1 <= jj <= t_2
-    ])
-    return [idxs[0], idxs[-1]]
-
-
-def _monthly_t(cubes):
-    """Rearrange time points for monthly data."""
-    # get original cubes tpoints
-    days = {day for cube in cubes for day in _datetime_to_int_days(cube)}
-    return sorted(days)
-
-
-def _full_time_slice(cubes, ndat, indices, ndatarr, t_idx):
-    """Construct a contiguous collection over time."""
-    for idx_cube, cube in enumerate(cubes):
-        # reset mask
-        ndat.mask = True
-        ndat[indices[idx_cube]] = cube.data
-        if np.ma.is_masked(cube.data):
-            ndat.mask[indices[idx_cube]] = cube.data.mask
-        else:
-            ndat.mask[indices[idx_cube]] = False
-        ndatarr[idx_cube] = ndat[t_idx]
-
-    # return time slice
-    return ndatarr
-
+    t_unit = _get_consistent_time_unit(cubes)
```

Comment:
It would be computationally more efficient to just return `cubes` at this point if no change to the calendars is needed.

Reply:
And also more reliable, because if no change is needed, e.g. resetting the bounds and then guessing them might unexpectedly change them

Reply:
That won't work, because the function will still regrid time for monthly and daily data, even if the calendars are the same. But we can do this after or as part of #744.

Reply:
Ok, let's leave this for later.